src/sentry/api/paginator.py

```diff
@@ -746,3 +748,69 @@ def get_result(self, limit=100, cursor=None):
             results = self.on_results(results)
 
         return CursorResult(results=results, next=next_cursor, prev=prev_cursor)
+
+
+class SnubaRequestPaginator:
+    def __init__(
+        self,
+        query,
+        dataset,
+        app_id,
+        tenant_ids,
+        order_by=None,
+        max_limit=MAX_LIMIT,
+        converter=None,
+        on_results=None,
+    ):
+        self.query = query
+        self.dataset = dataset
+        self.app_id = app_id
+        self.tenant_ids = tenant_ids
+        self.desc = False
+        if order_by.startswith("-"):
+            self.desc = True
+            self.query = self.query.set_orderby([OrderBy(Column(order_by[1:]), Direction.DESC)])
+        else:
+            self.desc = False
+            self.query = self.query.set_orderby([OrderBy(Column(order_by), Direction.ASC)])
+        self.max_limit = max_limit
+        self.converter = converter
+        self.on_results = on_results
+
+    def build_next_snuba_request(self, cursor_value, query):
+        if cursor_value != 0:
+            query = query.set_offset(cursor_value)
+        return query
+
+    def get_result(self, limit=100, cursor=None):
+        if cursor is None:
+            cursor = Cursor(0, 0, 0)
+
+        limit = min(limit, self.max_limit)
+        self.query = self.query.set_limit(
+            limit + 1
+        )  # +1 to limit so that we can tell if there are more results left after the current page
+
+        request = Request(
+            dataset=self.dataset,
+            app_id=self.app_id,
+            query=self.build_next_snuba_request(cursor.value, self.query),
+            tenant_ids=self.tenant_ids,
+        )
+
+        results = raw_snql_query(request, referrer=self.app_id)["data"]
```

Comment:
in the tests (on local, haven't noticed failures from ci yet), sometimes this returns nothing or more events than expected. i imagine this has to do with the events not being saved to snuba right away or the tests being run in parallel to each other, as when the tests are run individually on local, they succeed.

Reply:
update: resolved!